### PR TITLE
Theme JSON: use block global style for block gap where no gap value exists in layout.php

### LINKED ISSUE #39789: Layout blockGap / spacing: Add support for gap set at the block level in theme.json and global styles
Following on from #39601 where we removed the blockGap / block spacing controls in global styles (since the gap value is not currently supported at the block level in global styles), this issue proposes a way to add it back in!

As of https://github.com/WordPress/gutenberg/pull/37360 where the CSS variable approach for `blockGap` was removed, we haven't had support for `blockGap` at the individual block level in global styles or `theme.json`. One possible path forward to re-instating it could be:

* Every block that uses the layout support is server-rendered and currently attempts to grab the block gap value from the block's attributes, and if it doesn't exist, falls back to using the global blockGap CSS variable.
* Can we add in logic in `layout.php` and in the JS rendering for the layout support so that if the block's attributes do contain a `blockGap` setting we then look-up `theme.json` to grab the `blockGap` value if one is set at the corresponding block name level?
* If this approach works, also roll it out to the Gallery block's custom code that implements gap support.

### PR DESCRIPTION
**Pull request by ramonjd**
## What?

Use block global style value (not the root) for block gap where no gap value exists. 

Resolves https://github.com/WordPress/gutenberg/issues/39789

## Why?

In https://github.com/WordPress/gutenberg/pull/37360 we removed `blockGap` from the block and elements arrays in `$theme_json`. 

The side-effect was that block gap values in global styles for blocks such as the Group block were neither saved in the editor or appear in global styles the frontend.

Because we were removing block gap global styles from the theme json tree it meant that we could never compile the styles and output them to the frontend.

As a temporary measure we've removed the block gap control UI from dimensions panel in https://github.com/WordPress/gutenberg/pull/39845.

This PR reinstates it.

## How?

- Only skips top-level style properties when compiling the styles. The block/element-level block gap values are not unset from the theme json tree.
- In layout.php, if there is no block-level value for blockGap, but a global styles value available for blockGap, use the latter.
- Reinstate the block gap control UI removed in https://github.com/WordPress/gutenberg/pull/39845

See: 
- https://github.com/WordPress/gutenberg/pull/39845#issuecomment-1081712594
- https://github.com/WordPress/gutenberg/issues/39789
- https://github.com/WordPress/gutenberg/pull/37360

## Testing Instruction
In the Site Editor, edit the global block gap value of a block that supports it, e.g., Group.

Check the editor and frontend to ensure the changes are visible. 

Ensure that no `top level` styles are overwritten, that is, any global block gap styles (`--wp--style--block-gap`)  set at the root level in the body tag should persist.

s

### CODE REVIEW

File: lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php [hunk 1]
```diff
@@ -119,6 +165,69 @@ protected function get_block_classes( $style_nodes ) {
...
+		$declarations = array();
+		if ( empty( $styles ) ) {
+			return $declarations;
+		}
+
+		foreach ( $properties as $css_property => $value_path ) {
+			// Some styles such as blockGap are only meant to be available at the top level (ROOT_BLOCK_SELECTOR),
+			// hence we only output styles at the top level.
+			if ( 'top' === _wp_array_get( self::VALID_STYLES, array( $value_path[0], $value_path[1] ), null ) && static::ROOT_BLOCK_SELECTOR !== $selector ) {
+				continue;
```
Comment (ramonjd) [inline lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php]: I'm suspect of this. It seems fragile, only in the sense that I'm not aware of any side-effects. 

Needs lots of testing.

**Comment by ramonjd** [comment]
I know the tests are failing. This PR is just to test out an idea from https://github.com/WordPress/gutenberg/issues/39789

File: lib/block-supports/layout.php [hunk 2]
```diff
@@ -36,9 +37,20 @@ function gutenberg_register_layout_support( $block_type ) {
...
  */
-function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support = false, $gap_value = null, $should_skip_gap_serialization = false ) {
+function gutenberg_get_layout_style( $block_name, $selector, $layout, $has_block_gap_support = false, $gap_value = null, $should_skip_gap_serialization = false ) {
 	$layout_type = isset( $layout['type'] ) ? $layout['type'] : 'default';
 
+	// If there is no block-level value for blockGap,
+	// but a global styles value available for blockGap,
+	// use the latter.
+	if ( $has_block_gap_support && empty( $gap_value ) ) {
+		$block_global_styles = gutenberg_get_global_styles( array( 'blocks', $block_name, 'spacing' ) );
```
Comment (andrewserong) [inline lib/block-supports/layout.php]: Just wondering, could we move this check to happen in `gutenberg_render_layout_support_flag` instead, before we parse / sanitize the`$gap_value` variable, so that by the time we're in this function, it's already the right gap value? That we might not need to update the signature for `gutenberg_get_layout_style`.

**Review by andrewserong — COMMENTED** [review]
This is awesome, thanks for picking this one up @ramonjd 🙇 — I haven't test it yet, but just added a quick drive-by comment about where we do the checking for the block-level value from global styles.

Looks like a good direction to go in to me, and happy to give it a more detailed test / look on Monday if no-one beats me to it!

File: lib/block-supports/layout.php (hunk 2 above)
Comment (ramonjd) [reply lib/block-supports/layout.php]: Good call 👍

File: lib/block-supports/layout.php [hunk 3]
```diff
@@ -162,6 +162,17 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
 
 	$class_name = wp_unique_id( 'wp-container-' );
 	$gap_value  = _wp_array_get( $block, array( 'attrs', 'style', 'spacing', 'blockGap' ) );
+
+	// If there is no block-level value for blockGap,
+	// but a global styles value available for blockGap,
+	// use the latter.
+	if ( empty( $gap_value ) ) {
```
Comment (youknowriad) [inline lib/block-supports/layout.php:180]: Why is this needed? isn't the fallback supposed to happen using the CSS variable?

Comment (youknowriad) [reply lib/block-supports/layout.php:180]: (And we can only set the CSS variable at the global level theme.json)

Comment (ramonjd) [reply lib/block-supports/layout.php:180]: > Why is this needed?

Good question. 

Logically it's not. 

I only put it there so we could skip over trying to fetch the global value via `gutenberg_get_global_styles` if a gap value set in the block's local styles attribute is found, which would take precedence anyway.

> isn't the fallback supposed to happen using the CSS variable? 

It will either way, unless I'm missing something, which could be the case 😇 

My thinking went along of these lines: if the block's local styles attribute does not contain a gap value, we'll try the block's global styles. 

Either could still be `null` by the time we call `gutenberg_get_layout_style()` in which case the fallback would be `--wp--style--block-gap`, which is set at the root. 

Does that sound right?

Comment (youknowriad) [reply lib/block-supports/layout.php:180]: It does feel like duplication to me, since if the first one is null, the fallback will be the CSS variable (which corresponds to the root value). 

Or in other words, if we do the fallback in php, why do we have the CSS variable in the first place.

Comment (ramonjd) [reply lib/block-supports/layout.php:180]: > It does feel like duplication to me, since if the first one is null, the fallback will be the CSS variable (which corresponds to the root value).

Thanks for clarifying.

Just so I understand your concerns, are you seeing somewhere where the root block gap value, as opposed to the block-level global style value, is returned from 

```php
$spacing_global_styles = gutenberg_get_global_styles( array( 'blocks', $block['blockName'], 'spacing' ) );
```

on this [line](https://github.com/WordPress/gutenberg/pull/39870/files#diff-324697e41855298e2f2c74b078f174e0cbc9075cef736ce9c1e2c169bf64652eR170)?

If so, yes, that would be a mistake.

The intention of this PR is to get gap working for block-level global styles by:

1. filtering out block-level global styles for blockGap, but NOT the root, in the style sheet. See [this line](https://github.com/WordPress/gutenberg/pull/39870/files#diff-c220112e642d42482d91af8a96dca5ab587adfa705140148f86c064bca995631R197)
2.  using the block-level global styles value, which still exists in the theme json tree, as a fallback ONLY IF the block attributes do not contain a `style.spacing.blockGap` value. If the former doesn't exist, we'll use the root value.

The effect of point 1 is that a user sets the value of, say, the Group block's gap in global styles, `.wp-block-group { --wp--style--block-gap: '{someCustomValue}px}` won't be rendered:

```html
<style id='global-styles-inline-css'>
    body {
            /* this will still be here! */
	    --wp--style--block-gap: 1.5rem;
    }
    /* block-level global styles rules will not be printed */
</style>
````

The consequence of point 2 – at least what I'm attempting - is that if a user has set the value of **Group block's gap** in global styles it will be used when an individual block's attributes do not contain a `style.spacing.blockGap` value.

So the priority should still be:

- try to use block attributes first
- then the block's global styles second
- finally the root value

If things aren't working that way in this PR it means I've done something wrong! 😬

Comment (youknowriad) [reply lib/block-supports/layout.php:180]: Oh really sorry @ramonjd I misunderstand the original purpose of this PR :) 

Now that I understand it better, I think it's actually a smart way to solve "gap" support in theme.json at all levels. 

That said, I'm still hesitant for a couple reasons:

 - This basically changes how the "style engine" of layout works, instead of being `styleObject => CSS`, it's now `( styleObject, theme.json ) => CSS`. It changes the mental model and will have an impact on the style engine work.
 - In #39281 we're planning to add "theme.json" like support to each block (containers), meaning a container a "group" block will be able to say: all social link blocks inside me should have a gap of 10px. Let's think a little bit about what that means. It means that we can have nested containers which each container defining a different block gap (or not) until we reach the root level. It means that to generate the right style here we need to be aware of all the parent blocks' theme.json. So instead of `( styleObject, theme.json ) => CSS`, it becomes `( styleObject, theme.jsonForAllMyContainers, rootTheme.json) => CSS`. I think it's probably very hard (if not impossible) to access the `theme.jsonForAllMyContainers` from a given block.

Comment (ramonjd) [reply lib/block-supports/layout.php:180]: You raise good points, thanks for clarifying. 

> This basically changes how the "style engine" of layout works, instead of being styleObject => CSS, it's now ( styleObject, theme.json ) => CSS. It changes the mental model and will have an impact on the style engine work.

Yeah, I see what you mean. We might run into trouble if we mix block support style attributes with global styles. 

> we're planning to add "theme.json" like support to each block (containers)

I wasn't aware of Style Partials (theme.json) so I agree that a change like the one in this PR might come back to bite us. 👍 

I was having a discussion with @tellthemachines about something similar over at [Try using variable for root padding value #39926](https://github.com/WordPress/gutenberg/pull/39926#discussion_r839253862)

I was wondering about an alternative: a `const VALID_TOP_LEVEL_STYLES` (or whatever) in `WP_Theme_JSON_Gutenberg` that explicitly defines unique, root-level-only CSS vars.

```
body {
    --wp--root--style--block-gap: 2px;
    --wp--root--style--padding: 10px;
     /* etc ... */
}
```

I haven't really thought this through, but I was venturing towards checking `VALID_TOP_LEVEL_STYLES` when we assemble the styles in `compute_style_properties` and appending `--wp--root-style--` for  `ROOT_BLOCK_SELECTOR` styles.

Or there might be a tricky way to assemble them in `sanitize()` via `$schema['styles']`, `$schema['styles']['elements']` and `$schema['styles']['blocks']`. 

Uniquely identifying global CSS vars has a benefit, aside from explicitly stating their purpose, in that their values can be used for top-level layout calculation. 

For example, in https://github.com/WordPress/gutenberg/pull/39926, full-width blocks can use a root padding CSS var set a negative margins so that they stretch to the "full-width" of the screen.

Ultimately, we need to know how to treat `blockGap` when it is a root or a block/element style value and output something different accordingly. 

I can chat about it with @andrewserong and see what we can cook up.

**Comment by ramonjd** [comment]
I've been toying around with alternatives, hoping to cover the requirements of [root padding CSS vars as well](https://github.com/WordPress/gutenberg/pull/39926/files#r841362562), but I can't yet find something that satisfies all conditions, namely:

1. Define a consistent gap across blocks and layouts (flex, flow...) at the root level (one of the reasons behind https://github.com/WordPress/gutenberg/pull/37360)
2. Reenable block theme.json/global style block gap support (we've disabled it in the UI, but it still doesn't function via theme.json)
3. Prioritize block theme.json/global style block gap values over theme.json/global style values

Stepping back, if the approach in this PR doesn't end up being the way to go, should we rethink how `blockGap` is serving us?

Ignoring how hard it would be to do, given that `blockGap` is closely coupled with our layout implementation, perhaps "spacing" is the wrong group for it, and it should be under "layout"?

1. `blockGap` doesn't work in practice like any of the other spacing properties (margin and padding).
2. The layout panel operates at the block-level (it doesn't appear in global styles). A `gap` or `spacingGap` or whatever  control would be part of the layout panel, and its value would overwrite the root value and that's it. Basically how it works right now.
3. `blockGap` having been taken over by the layout implementation, it could free us up to reintroduce `gap` block support for non-layout blocks, e.g., navigation, social icons, etc

I'm not sure, just thinking aloud right now 🔊

File: lib/block-supports/layout.php (hunk 3 above)
Comment (andrewserong) [reply lib/block-supports/layout.php:180]: Great questions.

> This basically changes how the "style engine" of layout works, instead of being styleObject => CSS, it's now ( styleObject, theme.json ) => CSS. It changes the mental model and will have an impact on the style engine work.

Totally, I think it reveals some pragmatic challenges surrounding how we output styles, particularly for the layout support, which as @ramonjd mentions further down, is a little different to the other block supports.

In the case of this PR (and the current implementation of the layout support), everything happens at the individual block level when it's rendered, so out of necessity, to factor in global styles, we've kinda got to "look up" to the parents in order to deal with those values (or use CSS variables to stand-in for it).

In an ideal circumstance, with the style engine, how differently might we be doing things? The model for the style engine is a lot clearer for supports like padding, margin, typography, etc, where we'd like to consolidate the approach at the individual block level and global styles, and using a block's classname is an appropriate hook for it. In some of the explorations (like https://github.com/WordPress/gutenberg/pull/39374) it looks like we'll be able to move more to global styles, but the unique block gap values was still a bit of mystery to me as to how best to handle it, and neatly declare values that get passed down / or cascade down correctly.

With block gap, because the value is used directly in different circumstances (rendering `flow` vs `flex` gap), it does create this additional challenge. I think because the use case is so different to each of the other block supports, I'm wondering if the approach in this PR would be an acceptable way to unblock getting the global styles support in, and if we could treat it as an implementation detail we could potentially swap out if and when partials in https://github.com/WordPress/gutenberg/issues/39281 is explored.

From my perspective, it could be worthwhile proceeding with this change, so long as:

* We're happy that folks will set `blockGap` values at the block level within `theme.json` and via the global styles interface in the site editor.
* So long as we're happy with the shape of the data as it's stored, then theoretically, we could change the implementation after the fact, once we have a clearer idea of how we'd like to handle outputting the layout styles in the style engine.

From some of the explorations we've looked at recently, given that a fair bit of the logic winds up being moved to the Theme JSON class anyway, I don't _think_ that the current change here would preclude further work, but very happy to be proved wrong, as I definitely don't want to nudge us into a decision that's difficult to reverse!

**Comment by andrewserong** [comment]
> Stepping back, if the approach in this PR doesn't end up being the way to go, should we rethink how blockGap is serving us?

I quite like the approach in this PR, but yes, it probably is worth giving some thought to how `blockGap` is working for us. One idea I had, which might not be at all viable, would be to re-use `blockGap` for blocks that don't opt-in to layout support, by implementing the `gap` value directly, but only if the block does not use the `layout` support. So something like:

* The block opts-in to both layout and blockGap — the blockGap gets treated as part of the layout support as it does now and in this PR
* The block opts-in to blockGap but _not_ layout — the blockGap gets treated as the `gap` CSS property in much the same way as the `padding` and `margin` block supports use it

Not sure how viable that'd be, but just an idea, so we don't wind up with two gap features that are very similar 🤔 

> Ignoring how hard it would be to do, given that blockGap is closely coupled with our layout implementation, perhaps "spacing" is the wrong group for it, and it should be under "layout"?

I actually quite like "spacing" as the grouping for it, because controlling the gap between things is so closely related to padding and margin, where other controls like orientation, justification, etc, aren't unit-based, so more naturally fit into `layout` from a UI perspective. So, keeping the structure of where we put the value close to where we display the input field, feels natural to me, even if it does mean that our logic is a little split in terms of what winds up outputting the value.

**Comment by ramonjd** [comment]
> would be to re-use blockGap for blocks that don't opt-in to layout support, by implementing the gap value directly, but only if the block does not use the layout support

Great stuff. I like this idea, at least as a soft alternative to keep the spirit of this PR alive. 🥇

File: lib/block-supports/layout.php (hunk 3 above)
Comment (youknowriad) [reply lib/block-supports/layout.php:180]: > With block gap, because the value is used directly in different circumstances (rendering flow vs flex gap), it does create this additional challenge. 

One alternative could be to keep the current logic (selectors specific styles but output something like):

```
.selector.is-layout-flex {
   gap: something;
}

.selector.is-layout-default > * {
   margin-top: something;
}
```

To be honest, I didn't think much about this, it's just a random idea for now, but it's an alternative that allows us to avoid that dependency to parent global style.

 > From some of the explorations we've looked at recently, given that a fair bit of the logic winds up being moved to the Theme JSON class anyway, I don't think that the current change here would preclude further work, but very happy to be proved wrong

I don't know 🤷 to be honest, personally I don't have a full confidence here but I trust your judgement.

**Comment by ramonjd** [comment]
> The block opts-in to blockGap but not layout — the blockGap gets treated as the gap CSS property in much the same way as the padding and margin block supports use it

I was thinking about this today. To make the gap CSS property useful, the block would have to employ flex anyway. 

It would be nice to separate the two still (layout gap and regular gap). 

> styleObject => CSS, it's now ( styleObject, theme.json ) => CSS. It changes the mental model and will have an impact on the style engine work.

I think it's gap that starkly highlights the edge case because of the way it works at the root level. 

I can't yet come up with any other use case for why we'd want to look up global styles at the block supports level, which means Gutenberg's  inheritance model is working fairly well I guess 😄 

As @andrewserong mentions [above](https://github.com/WordPress/gutenberg/pull/39870#discussion_r841413547), the other block supports play much more nicely.

So in my mind, either we accept that block gap is a special case and carry on, or we split/change the way block gap works (no idea about that one 🤣 ) or we run with this PR, and look up the Gutenberg style inheritance tree to determine the fallback. The latter would fix the bug until something better comes along.

To be honest I'm not sure either.

Given the the inheritance model of global styles > block global styles > block styles, and somewhere in between elements and, soonish, style "partials" I wonder how long we can get away with the Style Engine only knowing about a block's style object to the exclusion of all other styles that may effect that particular block.

Maybe things will become clearer once we work out how the Style Engine will gather and output block styles/classnames: we'll be able to see inheritance/cascade on the frontend. The answer might lie in the way we structure the CSS as opposed to PHP logic.

🤷

File: lib/block-supports/layout.php (hunk 3 above)
Comment (andrewserong) [reply lib/block-supports/layout.php:180]: > One alternative could be to keep the current logic (selectors specific styles but output something like):

I really like that approach — I'd like it if we can output something like that in the longer term. I might be overthinking it, but I think that approach slightly opens up the can-of-worms surrounding rendering out semantic classnames, which we're in more of the proposal / exploratory phase of (since we'd need to adjust the logic for how the theme JSON class outputs styles, and then the layout support to inject the classnames).

> I don't know 🤷 to be honest, personally I don't have a full confidence here but I trust your judgement.

Thank you, much appreciated! Since I was one of the folks to propose the idea of looking up global styles, I think I'd feel more comfortable with getting a couple more opinions if you're not totally sold on the idea, to confidence check before proceeding. I think my main interest is to see how viable we think a short-term fix is (so that we can fix this bug in WP 6.0) and then aim to have the class-based approach like in your code snippet in time for WP 6.1.

**Comment by andrewserong** [comment]
> To make the gap CSS property useful, the block would have to employ flex anyway.

Or grid! In that case, we're really saying that the block that's opted-in has provided its own `display` styles, I suppose.

> or we run with this PR, and look up the Gutenberg style inheritance tree to determine the fallback. The latter would fix the bug until something better comes along.

I think I'm leaning in this direction, personally. Given that in alternate ways of doing it (like rendering out the class name from the theme JSON class) we're still pulling the value from the same place in the data structure, it _seems_ like the proposed fix here isn't one that would be too difficult to revert if we needed to, since it doesn't alter any saved data.

> Given the the inheritance model of global styles > block global styles > block styles, and somewhere in between elements and, soonish, style "partials" I wonder how long we can get away with the Style Engine only knowing about a block's style object to the exclusion of all other styles that may effect that particular block.

That's a good question 🤔, it's very much from the perspective of everything cascading down correctly, which I really like. 🤞 between being able to add more specific selectors, multiple attributes per classname, etc, and consolidating rendering a style object between global styles + at the individual block level, I think we should be able to cover most use cases. Riad's [very cool suggestion for the classname-based approach above](https://github.com/WordPress/gutenberg/pull/39870#discussion_r841631622) is a good example of one of the things that should be fairly straightforward if and when we settle on one of the layout-as-presets approaches.

It's highly likely we'll still encounter edge cases that don't _quite_ fit into things, but at least from the style engine perspective — if we've simplified and consolidated rendering a fair bit, and removed a lot of the duplication and one-off code in block supports, _and_ if folks find it easy enough to contribute to, then I think we'll have succeeded.

> Maybe things will become clearer once we work out how the Style Engine will gather and output block styles/classnames: we'll be able to see inheritance/cascade on the frontend.

I think so too 🙂

**Comment by ramonjd** [comment]
> I think I'm leaning in this direction, personally. Given that in alternate ways of doing it (like rendering out the class name from the theme JSON class) we're still pulling the value from the same place in the data structure, it seems like the proposed fix here isn't one that would be too difficult to revert if we needed to, since it doesn't alter any saved data.

Thanks for your thoughts @andrewserong 🙇 

One thing that only occurred to me just now is that `layout.php` could be classified as a bit of an outlier too. It introduces some intricate styles that require unique logic. 

`layout.php` is also looking at global properties already: 

```php
$block_gap             = gutenberg_get_global_settings( array( 'spacing', 'blockGap' ) );
$default_layout       = gutenberg_get_global_settings( array( 'layout' ) );
```

Maybe part of the paradigm should be to consider it a case of "layout implementation exceptionalism". 😄 

Seeing it this way, it makes me sleep better ~at my desk~ at night when comtemplating the way we've approached things here as well.

File: lib/block-supports/layout.php (hunk 3 above)
Comment (youknowriad) [reply lib/block-supports/layout.php:180]: cc @jorgefilipecosta @oandregal @mcsf if you have any opinions here.

**Comment by ndiego** [comment]
@ramonjd  is this still something we are trying to land in 6.0. If not, I will remove from the [project board](https://github.com/orgs/WordPress/projects/27/views/1). Thanks!

**Comment by ramonjd** [comment]
> @ramonjd is this still something we are trying to land in 6.0. If not, I will remove from the [project board](https://github.com/orgs/WordPress/projects/27/views/1). Thanks!

Thanks for checking @ndiego 

I'll remove it. I'm not convinced we'll come up with a satisfactory compromise in time.

**Comment by scruffian** [comment]
This work seems to be connected to the idea of block partials, which is blocked by conditional loading of the styles for each blocks. I had a go at introducing that change here: https://github.com/WordPress/gutenberg/pull/40513

I'd appreciate some input from you on this change as I think it will help to drive forward the changes here...

**Comment by ramonjd** [comment]
Noting that there is a WIP alternative to this PR in the works:

- https://github.com/WordPress/gutenberg/pull/40875

So I might put this one on ice for a bit.

**Comment by andrewserong** [comment]
Thanks Ramon! Yes, I'm currently hacking away at #40875 to explore Riad's suggestion from this comment (https://github.com/WordPress/gutenberg/pull/39870#discussion_r841631622) to see if it could be a viable alternative. The PR's a little broken right now but I'll continue playing around with it 🙂

**Comment by ramonjd** [comment]
Closing in favour of https://github.com/WordPress/gutenberg/pull/40875
